File: as/src/main/java/com/demo/mjpeg/MjpegView.java
// http://stackoverflow.com/questions/10550139/android-ics-and-mjpeg-using-asynctask
package com.demo.mjpeg;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffXfermode;
import android.graphics.Rect;
import android.graphics.Typeface;
import android.os.Build;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.util.Log;
import android.view.SurfaceHolder;
import android.view.SurfaceView;

import org.apache.commons.io.input.BoundedInputStream;

import java.io.IOException;
import java.util.Locale;

public class MjpegView extends SurfaceView implements SurfaceHolder.Callback {
    private static final String TAG = "MjpegView";
    private MjpegViewThread thread;
    private MjpegInputStream mIn;
    private boolean mRun;
    private boolean surfaceDone;
    private final Paint fpsTextPaint = new Paint();
    private int dispWidth;
    private int dispHeight;

    public MjpegView(Context context) {
        super(context);
        init(context);
    }

    public MjpegView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    private void init(Context context) {
        SurfaceHolder holder = getHolder();
        holder.addCallback(this);
        thread = new MjpegViewThread(holder);
        setFocusable(true);
        fpsTextPaint.setTextAlign(Paint.Align.RIGHT);
        fpsTextPaint.setTextSize(12);
        fpsTextPaint.setTypeface(Typeface.DEFAULT);
        fpsTextPaint.setColor(Color.WHITE);
        dispWidth = getWidth();
        dispHeight = getHeight();
    }


    public void setSource(MjpegInputStream source) {
        mIn = source;
    }

    public synchronized void startPlayback() {
        startPlaybackInternal();
    }

    private void startPlaybackInternal() {
        if (mIn != null) {
            mRun = true;
            thread.start();
        }
    }

    public synchronized void stopPlayback() {
        stopPlaybackInternal();
    }

    private void stopPlaybackInternal() {
        if (mRun) {
            mRun = false;
            thread.join();
        }
    }

    @Override
    public void surfaceChanged(SurfaceHolder holder, int f, int w, int h) {
        thread.setSurfaceSize(w, h);
    }

    @Override
    public void surfaceCreated(SurfaceHolder holder) {
        surfaceDone = true;
    }

    @Override
    public void surfaceDestroyed(SurfaceHolder holder) {
        surfaceDone = false;
        stopPlayback();
    }

    class MjpegViewThread {
        private final SurfaceHolder mSurfaceHolder;
        @Nullable
        private Thread thread;

        public MjpegViewThread(SurfaceHolder surfaceHolder) {
            mSurfaceHolder = surfaceHolder;
        }

        private void initThread() {
            join();
            if (thread != null)
                join();
            thread = new MjpegRenderThread(this.mSurfaceHolder);
        }

        public void join() {
            if (thread != null) {
                boolean retry = true;
                while (retry) {
                    try {
                        thread.join(3000, 0);
                        retry = false;
                    } catch (InterruptedException e) {
                        Log.e(this.getClass().getSimpleName(), Log.getStackTraceString(e));
                    }
                }
                thread = null;
            }
        }

        public void setSurfaceSize(int width, int height) {
            synchronized (mSurfaceHolder) {
                dispWidth = width;
                dispHeight = height;
            }
        }

        public void start() {
            initThread();
            if (thread != null)
                thread.start();
        }

    }

    /// TODO: rewrite from scratch. needs redesign.
    private class MjpegRenderThread extends Thread {
        private final SurfaceHolder mSurfaceHolder;
        private int mFrameCounter;
        private long mStart;

        @Nullable
        private Bitmap mBitmap;
        @Nullable
        private BoundedInputStream mFrame;
        @NonNull
        private final byte[] mTempStorage = new byte[100000];
        private final PorterDuffXfermode mMode = new PorterDuffXfermode(PorterDuff.Mode.DST_OVER);
        private String mFpsStr = "";

        public MjpegRenderThread(SurfaceHolder holder) {
            this.mSurfaceHolder = holder;
        }

        @Override
        public void run() {
            mStart = System.currentTimeMillis();

            while (mRun) {
                if (!surfaceDone)
                    continue;
                Canvas canvas = null;
                try {
                    mFrame = mIn.readMjpegFrame();
                    if (mFrame == null)
                        continue;

                    Rect destRect = ExtractNextBitmap();
                    if (destRect == null)
                        continue;


                    if (null == (canvas = mSurfaceHolder.lockCanvas()))
                        continue;

                    if (mBitmap != null)
                        DrawToCanvas(canvas, destRect);

                } catch (IOException e) {
                    mRun = false;
                } finally {

                    if (canvas != null) {
                        mSurfaceHolder.unlockCanvasAndPost(canvas);
                        canvas = null;
                    }
                    if (mFrame != null)
                        try {
                            // Stream is bound to mBitmap, but we do not need
                            // both of them later, mBitmap to be reused.
                            mFrame.close();
                        } catch (IOException e) {
                            mRun = false;
                        }
                    mFrame = null;
                }
            }
        }

        synchronized private Rect ExtractNextBitmap() {

            BitmapFactory.Options opts = new BitmapFactory.Options();
//            opts.inJustDecodeBounds = true;
//            try {
//                mFrame.mark(mFrame.available());
//                BitmapFactory.decodeStream(mFrame, null, opts);
//                mFrame.reset();
//            } catch (IOException e) {
//                return null;
//            }


//            opts.inSampleSize = 1;
//            while (destRect.width() < opts.outWidth >> 1 && destRect.height() < opts.outHeight >> 1) {
//                opts.inSampleSize <<= 1;
//                opts.outWidth >>= 1;
//                opts.outHeight >>= 1;
//            }
//
//            opts.inScaled = true;

            //final int scale = (int) getResources().getDisplayMetrics().density;
            //opts.inTargetDensity = scale * opts.inSampleSize;
            //opts.inScreenDensity = scale;
            //opts.inDensity = scale * opts.outHeight / destRect.height();

            opts.inBitmap = mBitmap; // reuse if possible
            opts.inMutable = true;

            opts.inPurgeable = true;
            opts.inInputShareable = true;
            //opts.inPreferredConfig = Bitmap.Config.RGB_565;
            opts.inTempStorage = mTempStorage;

            Bitmap bm;

            try {
                //opts.inJustDecodeBounds = false;
                bm = BitmapFactory.decodeStream(mFrame, null, opts);
            } catch (IllegalArgumentException e) {
                bm = null;
            }

            if (bm == null)
                return null;

            if (mBitmap != null && bm != mBitmap) {
                // Was not reused
                mBitmap.recycle();
                Log.v(TAG, "Bitmap was not reused, recycled.");
            }

            mBitmap = bm;
            //mBitmap.setDensity(Bitmap.DENSITY_NONE);
            return destRect(bm.getWidth(), bm.getHeight());
        }

        private void DrawToCanvas(@NonNull Canvas canvas, @NonNull Rect destRect) {

            mFrameCounter++;
            final long now = System.currentTimeMillis();
            final long elapsedMs = now - mStart;
            final int TIME_FRAME_MS = 5000;
            if (elapsedMs >= TIME_FRAME_MS) {
                Log.v(TAG, "elapsed " + elapsedMs);
                mStart = now;
                float fps = 1000.0f * mFrameCounter / TIME_FRAME_MS;
                mFrameCounter = 0;
                mFpsStr = String.format(Locale.getDefault(), "%.1f", fps);
            }

            synchronized (mSurfaceHolder) {
                canvas.drawColor(Color.BLACK);
                synchronized (this) {
                    if (mBitmap != null)
                        canvas.drawBitmap(mBitmap, null, destRect, null);
                }
                canvas.drawText(mFpsStr, dispWidth - 1, -fpsTextPaint.ascent(), fpsTextPaint);
            }


        }


        @NonNull
        private Rect destRect(int bmw, int bmh) {
            int tempX;
            int tempY;
            float bmasp = (float) bmw / (float) bmh;
            bmw = dispWidth;
            bmh = (int) (dispWidth / bmasp);
            if (bmh > dispHeight) {
                bmh = dispHeight;
                bmw = (int) (dispHeight * bmasp);
            }
            tempX = dispWidth / 2 - bmw / 2;
            tempY = dispHeight / 2 - bmh / 2;
            return new Rect(tempX, tempY, bmw + tempX, bmh + tempY);
        }

    }


}

File: as/app/src/main/kotlin/com/demo/mjpeg/MjpegView.kt
// http://stackoverflow.com/questions/10550139/android-ics-and-mjpeg-using-asynctask
package com.demo.mjpeg

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.PorterDuff
import android.graphics.PorterDuffXfermode
import android.graphics.Rect
import android.graphics.Typeface
import android.util.AttributeSet
import android.util.Log
import android.view.SurfaceHolder
import android.view.SurfaceView

import java.io.IOException

class MjpegView : SurfaceView, SurfaceHolder.Callback {
    inner class MjpegViewThread(private val mSurfaceHolder: SurfaceHolder) {
        private var thread: Thread? = null
        private var frameCounter: Int = 0
        private var start: Long = 0
        private var ovl: Bitmap? = null

        private fun destRect(bmw_: Int, bmh_: Int): Rect? {
            val tempx: Int
            val tempy: Int
            if (displayMode == MjpegView.SIZE_BEST_FIT) {
                val bmasp = bmw_.toFloat() / bmh_.toFloat()
                var bmw = dispWidth
                var bmh = (dispWidth / bmasp).toInt()
                if (bmh > dispHeight) {
                    bmh = dispHeight
                    bmw = (dispHeight * bmasp).toInt()
                }
                tempx = dispWidth / 2 - bmw / 2
                tempy = dispHeight / 2 - bmh / 2
                return Rect(tempx, tempy, bmw + tempx, bmh + tempy)
            }
            return null
        }

        private fun initThread() {
            join()
            thread = object : Thread() {
                override fun run() {
                    start = System.currentTimeMillis()
                    val mode = PorterDuffXfermode(PorterDuff.Mode.DST_OVER)

                    val p = Paint()
                    while (mRun) {
                        if (surfaceDone) {
                            val c = mSurfaceHolder.lockCanvas()
                            try {
                                synchronized (mSurfaceHolder) {
                                    try {
                                        val bm = mIn!!.readMjpegFrame()

                                        val destRect = destRect(bm.width, bm.height)

                                        if (destRect != null) {

                                            c!!.drawColor(Color.BLACK)
                                            c.drawBitmap(bm, null, destRect, p)
                                            if (showFps) {
                                                p.setXfermode(mode)
                                                if (ovl != null) {
                                                    val height = if ((ovlPos and 1) == 1)
                                                        destRect.top
                                                    else
                                                        destRect.bottom - ovl!!.height
                                                    val width = if ((ovlPos and 8) == 8)
                                                        destRect.left
                                                    else
                                                        destRect.right - ovl!!.width
                                                    c.drawBitmap(ovl, width.toFloat(), height.toFloat(), null)
                                                }
                                                p.setXfermode(null)
                                                frameCounter++
                                                if (System.currentTimeMillis() - start >= 1000) {
                                                    val fps = frameCounter.toString() + " fps"
                                                    frameCounter = 0
                                                    start = System.currentTimeMillis()
                                                    ovl = makeFpsOverlay(overlayPaint!!, fps)
                                                }
                                            }
                                        }
                                    } catch (e: IOException) {
                                        e.getStackTrace()
                                        Log.d(TAG, "catch IOException hit in run", e)
                                    }

                                }
                            } finally {
                                if (c != null) {
                                    mSurfaceHolder.unlockCanvasAndPost(c)
                                }
                            }
                        }
                    }
                }
            }
        }

        fun join() {
            if (thread != null) {
                var retry = true
                while (retry) {
                    try {
                        thread!!.join(3000, 0)
                        retry = false
                    } catch (e: InterruptedException) {
                        Log.e(this.javaClass.simpleName, Log.getStackTraceString(e))
                    }

                }
                thread = null
            }
        }

        private fun makeFpsOverlay(p: Paint, text: String): Bitmap {
            val b = Rect()
            p.getTextBounds(text, 0, text.length(), b)
            val bWidth = b.width() + 2
            val bHeight = b.height() + 2
            val bm = Bitmap.createBitmap(bWidth, bHeight, Bitmap.Config.ARGB_8888)
            val c = Canvas(bm)
            p.color = overlayBackgroundColor
            c.drawRect(0f, 0f, bWidth.toFloat(), bHeight.toFloat(), p)
            p.color = overlayTextColor
            c.drawText(text, (-b.left + 1).toFloat(), bHeight / 2 - (p.ascent() + p.descent()) / 2 + 1, p)
            return bm
        }

        fun setSurfaceSize(width: Int, height: Int) {
            synchronized (mSurfaceHolder) {
                dispWidth = width
                dispHeight = height
            }
        }

        fun start() {
            initThread()
            thread!!.start()
        }
    }

    private var thread: MjpegViewThread? = null
    private var mIn: MjpegInputStream? = null
    private var showFps: Boolean = false
    private var mRun: Boolean = false
    private var surfaceDone: Boolean = false
    private var overlayPaint: Paint? = null
    private var overlayTextColor: Int = 0
    private var overlayBackgroundColor: Int = 0
    private var ovlPos: Int = 0
    private var dispWidth: Int = 0
    private var dispHeight: Int = 0

    private var displayMode: Int = SIZE_BEST_FIT

    constructor(context: Context) : super(context) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init()
    }

    private fun init() {
        val holder = holder
        holder.addCallback(this)
        thread = MjpegViewThread(holder)
        isFocusable = true
        overlayPaint = Paint()
        overlayPaint!!.textAlign = Paint.Align.LEFT
        overlayPaint!!.textSize = 12f
        overlayPaint!!.setTypeface(Typeface.DEFAULT)
        overlayTextColor = Color.WHITE
        overlayBackgroundColor = Color.BLACK
        ovlPos = MjpegView.POSITION_LOWER_RIGHT
        dispWidth = width
        dispHeight = height
    }


    fun setOverlayPosition(p: Int) {
        ovlPos = p
    }


    fun setSource(source: MjpegInputStream?) {
        mIn = source
        // startPlayback();
    }

    fun showFps(b: Boolean) {
        showFps = b
    }

    fun startPlayback() {
        if (mIn != null) {
            mRun = true
            thread!!.start()
        }
    }

    fun stopPlayback() {
        mRun = false
        thread!!.join()
    }

    override fun surfaceChanged(holder: SurfaceHolder, f: Int, w: Int, h: Int) {
        thread!!.setSurfaceSize(w, h)
    }

    override fun surfaceCreated(holder: SurfaceHolder) {
        surfaceDone = true
    }

    override fun surfaceDestroyed(holder: SurfaceHolder) {
        surfaceDone = false
        stopPlayback()
    }

    companion object {

        private val TAG = "MjpegView"
        const val POSITION_UPPER_RIGHT = 3

        const val POSITION_LOWER_RIGHT = 6
        const val SIZE_BEST_FIT = 4


    }
}